File: reciter.py
#!/usr/bin/python
# Copyright (C) 2009  David Roberts <d@vidr.cc>
#
# This program is free software; you can redistribute it and/or
# modify it under the terms of the GNU General Public License
# as published by the Free Software Foundation; either version 2
# of the License, or (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program; if not, write to the Free Software
# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
# 02110-1301, USA.

import sys

"""
Converts English text to a string of SAMPA phonemes.
$ python reciter.py The quick brown fox jumps over the lazy dog.
D@ kwIk br@Un fQks dZVmps @Uv@R D@ l{zI dQg
"""

# Regular English Pronunciation rules
# <http://www.phon.ucl.ac.uk/home/mark/regeng/>
rules = [
    "#|BE+||bI",
    "#|RE+||rI",
    "|-ABLE|#|@b@l",
    "|-AGE||IdZ",
    "|-AL|#|@l",
    "|-ARY||@rI",
    "|ATION|#|eIS@n",
    "|-ED|#|-d",
    "|-ES|#|-s",
    "|-EST|#|Ist",
    "|-FUL|#|fUl",
    "L#|LY|#|I",
    "#|LY|#|lI",
    "|-TY|#|tI",
    "#|MENT|#|m@nt",
    "|-N'T|#|@nt",
    "|-S|#|-s",
    "|-'S||-s",
    "|-S'|#|-s",
    "|SSION|#|S@n",
    "|SION|#|Z@n",
    "|TION|#|S@n",
    "|-Y|#|I",
    "#|A|#|eI",
    "#|ARE|#|AR",
    "#|AS|#|{z",
    "#|B|#|bi",
    "#|C|#|si",
    "#|COULD|#|kUd",
    "#|D|#|di",
    "#|DO|#|du",
    "#|DOES|#|dVz",
    "#|DON'T|#|d@Unt",
    "#|E|#|i",
    "#|F|#|ef",
    "#|G|#|dZi",
    "#|H|#|eItS",
    "#|HAS|#|h{z",
    "#|HAVE|#|h{v",
    "#|HER|#|h3R",
    "#|HIS|#|hIz",
    "#|HOW|#|haU",
    "#|I|#|aI",
    "#|IS|#|Iz",
    "#|J|#|dZeI",
    "#|K|#|keI",
    "#|L|#|el",
    "#|M|#|em",
    "#|N|#|en",
    "#|O|#|@U",
    "#|OF|#|Qv",
    "#|ONE|#|wVn",
    "#|P|#|pi",
    "#|Q|#|kju",
    "#|R|#|AR",
    "#|S|#|es",
    "#|SHE|#|Si",
    "#|SHOULD|#|SUd",
    "#|SOME|#|sVm",
    "#|T|#|ti",
    "#|THAN|#|D{n",
    "#|THAT|#|D{t",
    "#|THE|#|D@",
    "#|THEIR|#|De@R",
    "#|THEM|#|Dem",
    "#|THEN|#|Den",
    "#|THERE|#|De@R",
    "#|THESE|#|Diz",
    "#|THEY|#|DeI",
    "#|THIS|#|DIs",
    "#|THOSE|#|D@Uz",
    "#|TO|#|tu",
    "#|TWO|#|tu",
    "#|U|#|ju",
    "#|V|#|vi",
    "#|W|#|dVb@lju",
    "#|WANT|#|wQnt",
    "#|WAS|#|wQz",
    "#|WHAT|#|wQt",
    "#|WHERE|#|we@R",
    "#|WHOLE|#|h@Ul",
    "#|WHO||hu",
    "#|WITH|#|wID",
    "#|WOULD|#|wUd",
    "#|X|#|eks",
    "#|Y|#|waI",
    "#|YOU|#|ju",
    "#|Z|#|zed",
    "|TCH||tS",
    "S|CH||k",
    "|CH|^|k",
    "|CH||tS",
    "#|WH||w",
    "#|KN||n",
    "^|I|GH|aI",
    "|TH|ER|D",
    "|TH||T",
    "|SH||S",
    "|DGE|#|dZ",
    "|GE|#|dZ",
    "|NG||N",
    "|NK||Nk",
    "|BB||b",
    "|DD||d",
    "|FF||f",
    "|A|LL|O",
    "|LL||l",
    "|MM||m",
    "|NN||n",
    "|PP||p",
    "|RR||r",
    "|SS||s",
    "|TT||t",
    "|ZZ||z",
    "Q|U||w",
    "|EAR||I@",
    "|OOR||OR",
    "|OUR||OR",
    "|ARE|#|e@R",
    "|EA|D|e",
    "|ER|#|@R",
    "W|OR||3R",
    "|URE|#|@R",
    "|AI||eI",
    "|AU||O",
    "|EA||i",
    "|EE||i",
    "|IA||I@",
    "|IE||i",
    "|IO||I@",
    "|OA||@U",
    "|OE||@U",
    "|OI||OI",
    "|OO||U",
    "|OU||aU",
    "|UA||U@",
    "|UE||u",
    "|UI||I",
    "|A|^E|eI",
    "|A|NGE|eI",
    "#^|E|#|i",
    "^|LE|#|@l",
    "^|E|#|",
    "|I|^E#|aI",
    "|O|^E|@U",
    "^|O|#|@U",
    "|U|^E|ju",
    "|AR||AR",
    "|AW|.|@w",
    "|AW||O",
    "|AY||eI",
    "|ER||3R",
    "|EW||u",
    "|IR||3R",
    "|OR||OR",
    "|OW||@U",
    "|OY||OI",
    "|UR||3R",
    "|UY||aI",
    "|B||b",
    "|C|E|s",
    "|SC|I|s",
    "|C|I|s",
    "|C|Y|s",
    "|CK||k",
    "|C||k",
    "|F||f",
    "|GG||g",
    "|G|Y|dZ",
    "|GH||",
    "|G||g",
    "|PH||f",
    "|H||h",
    "|J||dZ",
    "|K||k",
    "|L||l",
    "|M||m",
    "|N||n",
    "|P||p",
    "|Q||k",
    "|T||t",
    "|V||v",
    "#|WR||R",
    "|W||w",
    "|X||ks",
    "^|Y|^|I",
    "#^^|Y|#|aI",
    "#^|Y|#|aI",
    "|Y|.|j",
    "|Y||I",
    "|Z||z",
    "|R||R",
    "|D||d",
    "|S||s",
    "^|A|#|@",
    "|A||{",
    "|E||e",
    "|I||I",
    "|O||Q",
    "|U||V",
]

def consonant(c):
    return c.upper() in 'BCDFGHJKLMNPQRSTVWXYZ'

def vowel(c):
    return c.upper() in 'AEIOUY'

def match_kind(kind, c):
    if kind == '#':   return c.upper() in '()#-+.'
    elif kind == '^': return consonant(c)
    elif kind == '.': return vowel(c)
    else:             return c.upper() == kind

def match_left(left, word):
    if len(word) < len(left): return False
    for i in xrange(len(left)):
        if not match_kind(left[-i-1], word[-i-1]):
            return False
    return True

def match_right(right, word):
    if len(word) < len(right): return False
    for i in xrange(len(right)):
        if not match_kind(right[i], word[i]):
            return False
    return True

def apply_rule(word, rule, output):
    left, central, right, phonemes = rule.split('|')
    match = -1
    while True:
        match = word.find(central, match+1)
        if match == -1: return word
        word_left = word[:match]
        word_right = word[match+len(central):]
        if not match_left(left, word_left): continue
        if not match_right(right, word_right): continue
        # rule matches
        output[match] = phonemes
        # convert matching characters to lowercase
        word = word_left + word[match:match+len(central)].lower() + word_right

def apply_plural_rule(phonemes):
    return phonemes.replace('s-s','sIz').replace('S-s','SIz') \
                   .replace('z-s','zIz').replace('Z-s','ZIz') \
                   .replace('p-s','ps') .replace('t-s','ts') \
                   .replace('k-s','ks') .replace('f-s','fs') \
                   .replace('T-s','Ts') .replace('-s','z')

def apply_past_tense_rule(phonemes):
    return phonemes.replace('t-d','tId').replace('d-d','dId') \
                   .replace('p-d','pt') .replace('k-d','kt') \
                   .replace('f-d','ft') .replace('T-d','Tt') \
                   .replace('s-d','st') .replace('S-d','St') \
                   .replace('-d','d')

def apply_r_deletion_rule(phonemes):
    return phonemes.replace('Ra','ra').replace('Re','re').replace('Ri','ri') \
                   .replace('Ro','ro').replace('Ru','ru').replace('RA','rA') \
                   .replace('RE','rE').replace('RI','rI').replace('RO','rO') \
                   .replace('RU','rU').replace('R3','r3').replace('R@','r@') \
                   .replace('RQ','rQ').replace('RV','rV').replace('R{','r{')

def main():
    phrase = ' '.join(sys.argv[1:]).upper().split()
    for word in phrase:
        word = '(' + word + ')'
        output = ['']*len(word)
        for rule in rules:
            word = apply_rule(word, rule, output)
        phonemes = ''.join(output)
        phonemes = apply_plural_rule(phonemes)
        phonemes = apply_past_tense_rule(phonemes)
        phonemes = apply_r_deletion_rule(phonemes)
        print phonemes,
    print
if __name__ == '__main__': main()

File: reciter.c
/* Copyright (C) 2009  David Roberts <d@vidr.cc>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 */

#include <stdio.h>

#include "cspeak.h"

/* Recite SAMPA phonemes read from stdin. */

const int FRAMES_CONSONANT = 6;
const int FRAMES_VOWEL = 8;
const int PITCH = 50;

#define CONSONANT if(phase==-1 || phase%2==1) phase++;
#define VOWEL if(phase==-1) phase = 1; \
              else if(phase%2==0) phase++; \
              if(phase>2) pitch = PITCH-2;

int peekchar(void) {
    int c = getchar();
    if(c != EOF) ungetc(c, stdin);
    return c;
}

int main() {
    char c;
    int phase = -1;
    int pitch = PITCH;
    
    while((c = getchar()) != EOF)
        switch(c) {
        case ' ': case '\n':
            synth(_NUL,FRAMES_VOWEL,pitch);
            phase = -1;
            pitch = PITCH;
            break;
        
        case 't': CONSONANT
                  if(peekchar() == 'S') {
                      synth(_CHa,FRAMES_CONSONANT,pitch);
                      synth(_CHb,FRAMES_CONSONANT,pitch);
                      getchar();
                  } else {
                      synth(_Ta,FRAMES_CONSONANT/3,pitch);
                      synth(_Tb,FRAMES_CONSONANT/3,pitch);
                      synth(_Tc,FRAMES_CONSONANT/3,pitch);
                  }
                  break;
        case 'd': CONSONANT
                  if(peekchar() == 'Z') {
                      synth(_Ja,FRAMES_CONSONANT/4,pitch);
                      synth(_Jb,FRAMES_CONSONANT/4,pitch);
                      synth(_Jc,FRAMES_CONSONANT/4,pitch);
                      synth(_Jd,FRAMES_CONSONANT/4,pitch);
                      getchar();
                  } else {
                      synth(_Da,FRAMES_CONSONANT/3,pitch);
                      synth(_Db,FRAMES_CONSONANT/3,pitch);
                      synth(_Dc,FRAMES_CONSONANT/3,pitch);
                  }
                  break;
        case 'p': CONSONANT
                  synth(_Pa,FRAMES_CONSONANT/3,pitch);
                  synth(_Pb,FRAMES_CONSONANT/3,pitch);
                  synth(_Pc,FRAMES_CONSONANT/3,pitch);
                  break;
        case 'b': CONSONANT
                  synth(_Ba,FRAMES_CONSONANT/3,pitch);
                  synth(_Bb,FRAMES_CONSONANT/3,pitch);
                  synth(_Bc,FRAMES_CONSONANT/3,pitch);
                  break;
        case 'k': CONSONANT
                  synth(_Ka,FRAMES_CONSONANT/3,pitch);
                  synth(_Kb,FRAMES_CONSONANT/3,pitch);
                  synth(_Kc,FRAMES_CONSONANT/3,pitch);
                  break;
        case 'g': CONSONANT
                  synth(_Ga,FRAMES_CONSONANT/3,pitch);
                  synth(_Gb,FRAMES_CONSONANT/3,pitch);
                  synth(_Gc,FRAMES_CONSONANT/3,pitch);
                  break;
        case 'f': CONSONANT synth(_F, FRAMES_CONSONANT,pitch); break;
        case 'v': CONSONANT synth(_V, FRAMES_CONSONANT,pitch); break;
        case 'T': CONSONANT synth(_TH,FRAMES_CONSONANT,pitch); break;
        case 'D': CONSONANT synth(_DH,FRAMES_CONSONANT,pitch); break;
        case 's': CONSONANT synth(_S, FRAMES_CONSONANT,pitch); break;
        case 'z': CONSONANT synth(_Z, FRAMES_CONSONANT,pitch); break;
        case 'S': CONSONANT synth(_SH,FRAMES_CONSONANT,pitch); break;
        case 'Z': CONSONANT synth(_ZH,FRAMES_CONSONANT,pitch); break;
        case 'h': CONSONANT synth(__H,FRAMES_CONSONANT,pitch); break;
        case 'm': CONSONANT synth(_M, FRAMES_CONSONANT,pitch); break;
        case 'n': CONSONANT synth(_N, FRAMES_CONSONANT,pitch); break;
        case 'N': CONSONANT synth(_NX,FRAMES_CONSONANT,pitch); break;
        case 'R': CONSONANT synth(_RX,FRAMES_CONSONANT,pitch); break;
        case 'r': CONSONANT synth(_R, FRAMES_CONSONANT,pitch); break;
        case 'l': CONSONANT synth(_L, FRAMES_CONSONANT,pitch); break;
        case 'w': CONSONANT synth(_W, FRAMES_CONSONANT,pitch); break;
        case 'j': CONSONANT synth(_Y, FRAMES_CONSONANT,pitch); break;
        
        case 'a': VOWEL
                  if(peekchar() == 'I') {
                      synth(_AY,FRAMES_VOWEL,pitch);
                      synth(_YX,FRAMES_CONSONANT,pitch);
                      getchar();
                  } else if(peekchar() == 'U') {
                      synth(_AW,FRAMES_VOWEL,pitch);
                      synth(_WX,FRAMES_CONSONANT,pitch);
                      getchar();
                  }
                  break;
        case 'e': VOWEL
                  if(peekchar() == 'I') {
                      synth(_EY,FRAMES_VOWEL,pitch);
                      synth(_YX,FRAMES_CONSONANT,pitch);
                      getchar();
                  } else {
                      synth(_EH,FRAMES_VOWEL,pitch);
                  }
                  break;
        case '@': VOWEL
                  if(peekchar() == 'U') {
                      synth(_OW,FRAMES_VOWEL,pitch);
                      synth(_WX,FRAMES_CONSONANT,pitch);
                      getchar();
                  } else {
                      synth(_ER,FRAMES_VOWEL,pitch);
                  }
                  break;
        case 'O': VOWEL
                  if(peekchar() == 'I') {
                      synth(_OY,FRAMES_VOWEL,pitch);
                      synth(_YX,FRAMES_CONSONANT,pitch);
                      getchar();
                  } else {
                      synth(_AO,FRAMES_VOWEL,pitch); break;
                  }
                  break;
        case 'I': VOWEL synth(_IH,FRAMES_VOWEL,pitch); break;
        case '{': VOWEL synth(_AE,FRAMES_VOWEL,pitch); break;
        case 'Q': VOWEL synth(_AA,FRAMES_VOWEL,pitch); break;
        case 'V': VOWEL synth(_AH,FRAMES_VOWEL,pitch); break;
        case 'U': VOWEL synth(_UH,FRAMES_VOWEL,pitch); break;
        case 'i': VOWEL synth(_IY,FRAMES_VOWEL,pitch); break;
        case 'u': VOWEL synth(_UX,FRAMES_VOWEL,pitch); break;
        case '3': VOWEL synth(_ER,FRAMES_VOWEL,pitch); break;
        case 'A': VOWEL synth(_AH,FRAMES_VOWEL,pitch); break;
        
        default: fprintf(stderr, "Unrecognised character: '%c'\n", c); break;
        }
    return 0;
}
